File: tests/test_task4_artifacts.py
"""Regressionstests für Konfigurationskonflikte und unvollständige Zwischenstände."""

import json
from pathlib import Path
import tempfile
import unittest

import pandas as pd

from src.task4_artifacts import (
    check_run_config, make_run_config, validate_parameter_table,
    validate_prediction_splits, write_run_config,
)


class ArtifactTests(unittest.TestCase):
    def test_changed_config_or_inputs_never_reuse_results(self):
        with tempfile.TemporaryDirectory() as directory:
            root = Path(directory)
            data = root / "splits.csv"
            data.write_text("split_id,inner_fold\n0,1\n")
            notebook = root / "model.ipynb"
            notebook.write_text(json.dumps({"cells": [{"source": ["model(C=C)"]}]}))
            def config(c_values):
                return make_run_config({"C_values": c_values}, [data], notebook, [0])
            original = config([0.01, 0.1, 1.0])
            manifest = root / "config.json"
            artifacts = [root / "predictions.csv", root / "models.csv"]
            self.assertFalse(check_run_config(manifest, original, artifacts, True))
            with self.assertRaises(FileNotFoundError):
                check_run_config(manifest, original, artifacts, False)
            for path in artifacts:
                path.write_text("existing results\n")
            with self.assertRaisesRegex(ValueError, "Konfigurationsnachweis fehlt"):
                check_run_config(manifest, original, artifacts, True)
            write_run_config(manifest, original)
            for training in [False, True]:
                self.assertTrue(check_run_config(manifest, original, artifacts, training))
                with self.assertRaisesRegex(ValueError, "Konfiguration"):
                    check_run_config(manifest, config([1.0]), artifacts, training)
            # Same donor names and seed, but a changed inner fold must be detected.
            data.write_text("split_id,inner_fold\n0,2\n")
            with self.assertRaisesRegex(ValueError, "inputs"):
                check_run_config(manifest, config([0.01, 0.1, 1.0]), artifacts, True)
            data.write_text("split_id,inner_fold\n0,1\n")
            notebook.write_text(json.dumps({"cells": [{"source": ["different_model(C=C)"]}]}))
            with self.assertRaisesRegex(ValueError, "implementation_sha256"):
                check_run_config(manifest, config([0.01, 0.1, 1.0]), artifacts, True)
            artifacts[1].unlink()
            with self.assertRaisesRegex(ValueError, "Unvollständiger Zwischenstand"):
                check_run_config(manifest, original, artifacts, True)

    def test_incomplete_or_misassigned_donor_predictions(self):
        splits = pd.DataFrame({
            "split_id": [0, 0], "split_seed": [7, 7], "donor_id": ["a", "b"],
            "label": [0, 1], "outer_partition": ["test", "test"],
        })
        predictions = splits.drop(columns="outer_partition").rename(columns={"label": "y_true"})
        predictions = predictions.assign(score=[0.2, 0.8], decision_threshold=0.5, y_pred=[0, 1])
        validate_prediction_splits(predictions, splits)
        for invalid in [
            predictions.iloc[:1],
            predictions.assign(split_seed=8),
            predictions.assign(y_true=[1, 0]),
            predictions.assign(y_pred=[1, 0]),
            predictions.assign(score=float("nan")),
        ]:
            with self.subTest(invalid=invalid.to_dict()):
                with self.assertRaises(ValueError):
                    validate_prediction_splits(invalid, splits)

    def test_missing_filter_or_marker_parameters(self):
        predictions = pd.DataFrame({"split_id": [0, 1]})
        parameters = pd.DataFrame({
            "split_id": [0, 0, 1, 1], "marker": ["CD3", "CD56"] * 2,
            "weight": [0.1, 0.2, 0.3, 0.4],
        })
        validate_parameter_table(parameters, predictions, ["CD3", "CD56"], ["split_id"])
        with self.assertRaisesRegex(ValueError, "Modellparameterspalten"):
            validate_parameter_table(parameters.drop(columns="weight"), predictions,
                                     ["CD3", "CD56"], ["split_id"], ["weight"])
        for invalid in [parameters.iloc[:-1], parameters.iloc[:2],
                        pd.concat([parameters, parameters.iloc[:1]]),
                        parameters.assign(weight=float("nan"))]:
            with self.assertRaises(ValueError):
                validate_parameter_table(invalid, predictions, ["CD3", "CD56"], ["split_id"])


if __name__ == "__main__":
    unittest.main()

File: tests/test_task4_artifacts.R
source("src/task4_artifacts.R")

expect_error <- function(expression) {
  failed <- tryCatch({force(expression); FALSE}, error = function(e) TRUE)
  stopifnot(failed)
}
directory <- tempfile("citrus_config_test_")
dir.create(directory)
config_path <- file.path(directory, "config.rds")
artifacts <- file.path(directory, c("predictions.csv", "selection.csv", "clusters.csv"))
config <- list(cofactor = 5, tolerance = 1e-10, inputs = c(splits = "original"))
stopifnot(!check_run_config(config_path, config, artifacts, TRUE))
expect_error(check_run_config(config_path, config, artifacts, FALSE))
invisible(file.create(artifacts))
expect_error(check_run_config(config_path, config, artifacts, TRUE))
saveRDS(config, config_path)
stopifnot(check_run_config(config_path, config, artifacts, TRUE))
for (training in c(TRUE, FALSE)) {
  changed <- config; changed$cofactor <- 10
  expect_error(check_run_config(config_path, changed, artifacts, training))
  changed <- config; changed$inputs <- c(splits = "changed inner folds")
  expect_error(check_run_config(config_path, changed, artifacts, training))
}
unlink(artifacts[3])
expect_error(check_run_config(config_path, config, artifacts, TRUE))

coefficients <- matrix(c(1, 0, 1e-15, 1e-10, -0.2), ncol = 1,
  dimnames = list(c("(Intercept)", "cluster 1 abundance", "cluster 2 abundance",
                    "cluster 3 abundance", "cluster 4 abundance"), NULL))
stopifnot(identical(effective_cluster_ids(coefficients, 1e-10), 4L))
selection <- data.frame(split_id = 0:2, selected_cluster_count = c(1, 1, 0))
predictions <- selection[rep(1:3, each = 6), ]
clusters <- data.frame(split_id = c(0, 1), cluster_id = c(1, 2),
                       marker = "CD3", coefficient = c(1e-15, -0.2))
clean <- normalize_cluster_tables(predictions, selection, clusters, 1e-10)
stopifnot(identical(clean$selection$selected_cluster_count, c(0L, 1L, 0L)),
          identical(clean$predictions$selected_cluster_count, rep(c(0L, 1L, 0L), each = 6)),
          nrow(clean$clusters) == 1L)
empty <- normalize_cluster_tables(predictions, selection, clusters[FALSE, ], 1e-10)
stopifnot(all(empty$predictions$selected_cluster_count == 0L))
unlink(directory, recursive = TRUE)
cat("Citrus-Konfiguration und Nulltoleranz: alle Tests bestanden.\n")
